examples/permissioned-token/assets/controller.py

```diff
@@ -0,0 +1,222 @@
+    var_total = Bytes("total_rules")
+    token_id = Bytes("token_id")
+    permissions_manager = Bytes("permissions_manager")	# permissions manager (asa.manager by default)
+    permission_id = Bytes("permissions_sc_app_id") # permissions smart contract application index
+
+    # Always verify that the RekeyTo property of any transaction is set to the ZeroAddress
+    # unless the contract is specifically involved in a rekeying operation.
+    no_rekey_addr = Txn.rekey_to() == Global.zero_address()
+
+    # retreive asset manager from Txn.ForeignAssets[0]
+    assetManager = AssetParam.manager(Int(0))
+
+    on_deployment = Seq([
+        assetManager, # load asset manager from store
+	Assert(And(
+		Txn.application_args.length() == Int(1),
+		# Txn.assets.length() == Int(1), [TEALv3]
+		no_rekey_addr,
+
+       	# Controller should be deployed by ASA.manager
+		assetManager.hasValue(),
+		Txn.sender() == assetManager.value()
+        )),
+
+        # total permission contracts during deploy is set to 0.
+        App.globalPut(var_total, Int(0)),
+
+        # set kill_status to false(0) during deploy
+        App.globalPut(var_is_killed, Int(0)),
+
+        # Save token index in controller's global state
+        App.globalPut(token_id, Btoi(Txn.application_args[0])), # or maybe use App.globalPut(token_id, Btoi(Txn.assets[0])) with TEALv3
+        Return(Int(1))
+    ])
+
+    # retreive asset reserve from Txn.ForeignAssets[0]
+    assetReserve = AssetParam.reserve(Int(0))
+    issue_token = Seq([
+        assetReserve,
+        Assert(And(
+		Txn.application_args.length() == Int(1),
+		# Txn.assets.length() == Int(1), [TEALv3]
+		Gtxn[1].type_enum() == TxnType.AssetTransfer,
+		Gtxn[1].xfer_asset() == App.globalGet(token_id), # verify if token index is correct
+
+		# Issue should only be done by token reserve
+		assetReserve.hasValue(),
+			Gtxn[0].sender() == assetReserve.value(),
+		Gtxn[1].asset_sender() == assetReserve.value(),
+
+		# only allow issue if token is not killed
+		App.globalGet(var_is_killed) == Int(0)
+        )),
+        Return(Int(1))
+    ])
+
+    # kill a token, after verifying transaction set is_killed to true(Int(1))
+    assetManager = AssetParam.manager(Int(0))
+    kill_token = Seq([
+        assetManager, # load asset_manager of Txn.ForeignAssets[0]
+        Assert(And(
+		Txn.application_args.length() == Int(1),
+		# Txn.assets.length() == Int(1), [TEALv3],
+		# Txn.assets[0] == App.globalGet(token_id), [TEALv3] # verify if token index is correct
+		Txn.type_enum() == TxnType.ApplicationCall,
+
+		# Only asset manager can kill the token
+		assetManager.hasValue(),
+		Txn.sender() == assetManager.value(),
+        )),
+
+        # finally set is_killed to true
+        App.globalPut(var_is_killed, Int(1)),
+        Return(Int(1))
+    ])
+
+    add_new_permission = Seq([
+        assetManager,
+        Assert(And(
+		Txn.application_args.length() == Int(3),
+		# Txn.assets[0] == App.globalGet(token_id), [TEALv3]
```

Comment:
It will give an invalid index error (if no args are passed & we try access Txn.application_args[1]). 
But line 95 is a basic check (if we're using `tx.args[0,1..n-1]` then we do an initial assert of `tx.args.length() == n`), and then we access it. It looks more intuitive. 
Let me know if you think we should remove.

Reply:
ok, we can keep it - let's add a comment what are the expected args.

Reply:
ok